src/vm/memory/segments.zig

```diff
@@ -168,12 +168,44 @@ pub const MemorySegmentManager = struct {
     /// # Returns
     ///
     /// An `AutoArrayHashMap` representing the computed effective sizes of memory segments.
-    pub fn computeEffectiveSize(self: *Self) !std.AutoArrayHashMap(u32, u32) {
-        for (self.memory.data.items, 0..) |segment, i| {
+    pub fn computeEffectiveSize(self: *Self, allow_tmp_segments: bool) !std.AutoArrayHashMap(i64, u32) {
+        if (self.segment_used_sizes.count() != 0) {
+            return self.segment_used_sizes;
+        }
+
+        // TODO: Check if memory is frozen. At the time of writting this function memory cannot be frozen so we cannot check if it frozen.
+        var first_segment_index: i64 = undefined;
+
+        if (allow_tmp_segments) {
+            first_segment_index = -@as(i64, self.memory.num_temp_segments);
+        } else {
+            first_segment_index = 0;
+        }
+
+        var first_index = first_segment_index;
+        while (first_index < self.memory.num_segments) {
+            try self.segment_used_sizes.put(first_segment_index, 0);
+            first_index += 1;
+        }
```

Comment:
For me you don't need this entire while stuff. I indeed see that it is done this way in the Python implementation but it is to initialize all sizes to 0 in case the segment contains nothing. For me the 0 will be automatically taken into account when you do `@intCast(segment.items.len)` below. In all cases this while is too expensive as it reviews all the segments while we do exactly the same loop afterwards.

Reply:
make sense, i was confused how it plays with the new structure so i left it there. will remove.